torch/csrc/jit/fusion_compiler.cpp

```diff
@@ -80,12 +84,297 @@ struct TensorInfo {
+    static const uint32 kPhiloxM4x32B = 0xCD9E8D57;
+
+    // Helper function to skip the next sample of 128-bits in the current stream.
+    PHILOX_DEVICE_INLINE void SkipOne() {
+      if (++counter_[0] == 0) {
+        if (++counter_[1] == 0) {
+          if (++counter_[2] == 0) {
+            ++counter_[3];
+          }
+        }
+      }
+    }
+
+    // Helper function to return the lower and higher 32-bits from two 32-bit
+    // integer multiplications.
+    PHILOX_DEVICE_INLINE
+    static void MultiplyHighLow(uint32 a, uint32 b, uint32* result_low,
+                                uint32* result_high) {
+  #ifndef __CUDA_ARCH__
+      const uint64 product = static_cast<uint64>(a) * b;
+      *result_low = static_cast<uint32>(product);
+      *result_high = static_cast<uint32>(product >> 32);
+  #else
+      *result_low = a * b;
+      *result_high = __umulhi(a, b);
+  #endif
+    }
+
+    // Helper function for a single round of the underlying Philox algorithm.
+    PHILOX_DEVICE_INLINE static ResultType ComputeSingleRound(
+        const ResultType& counter, const Key& key) {
+      uint32 lo0;
+      uint32 hi0;
+      MultiplyHighLow(kPhiloxM4x32A, counter[0], &lo0, &hi0);
+
+      uint32 lo1;
+      uint32 hi1;
+      MultiplyHighLow(kPhiloxM4x32B, counter[2], &lo1, &hi1);
+
+      ResultType result;
+      result[0] = hi1 ^ counter[1] ^ key[0];
+      result[1] = lo1;
+      result[2] = hi0 ^ counter[3] ^ key[1];
+      result[3] = lo0;
+      return result;
+    }
+
+    PHILOX_DEVICE_INLINE void RaiseKey(Key* key) {
+      (*key)[0] += kPhiloxW32A;
+      (*key)[1] += kPhiloxW32B;
+    }
+
+   private:
+    ResultType counter_;
+    Key key_;
+  };
+
+  // A Wrapper for Philox to get one random number at a time
+  class PhiloxWrapper {
+  public:
+    PHILOX_DEVICE_INLINE
+    explicit PhiloxWrapper(uint64 seed_lo, uint64 seed_hi, uint64 offset) {
+      philox = PhiloxRandom(seed_lo, seed_hi);
+      philox.Skip(offset / 4);
+      i = 0;
+    }
+    PHILOX_DEVICE_INLINE float operator()() {
+      if(i == 0) buf = philox();
+      uint32 ret = buf[i];
+      i = (i + 1) % 4;
+      static uint32 FLOAT_MASK = (1 << 24) - 1;
+      static float FLOAT_DIVISOR = 1.0f / (1 << 24);
+      return (ret & FLOAT_MASK) * FLOAT_DIVISOR;
+    }
+  private:
+    PhiloxRandom philox;
+    using ResultType = Array<uint32, 4>;
+    ResultType buf;
+    int i;
+  };
+
+  // }  // namespace random
+  // }  // namespace tensorflow
+
+  #endif  // TENSORFLOW_LIB_RANDOM_PHILOX_RANDOM_H_
+)";
+
+constexpr auto rand_param = ", uint64 seed, uint64 offset";
+constexpr auto rand_init = R"(
+  int idx = blockIdx.x*blockDim.x + threadIdx.x;
+  PhiloxWrapper rnd(seed, idx, offset);
+)";
 auto cuda_compilation_unit_template = CodeTemplate(R"(
 ${type_declarations}
 
 extern "C" __global__
-void ${kernelName}(IndexType totalElements, ${formals}) {
+void ${kernelName}(IndexType totalElements, ${formals} ${RandParam}) {
```

Comment:
Nit: our template engine supports automatically inserting the comma if the parameter is non-null, using the syntax `${,RandParam}`. So you don't need to put the comma in `rand_param` above.